apps/src/gamelab/animationListModule.js

```diff
@@ -0,0 +1,557 @@
+ *  dimensions and frame dimensions, framerate, name, last save time, version IDs,
+ *  URL to fetch the animation from the API, etc.
+ *
+ *  We serialize a smaller set of information.
+ *
+ *  We need to do a migration if the old style gets loaded.
+ *  See setInitialAnimationMetadata for how this works.
+ */
+import _ from 'lodash';
+import {combineReducers} from 'redux';
+import utils from '../utils';
+import {animations as animationsApi} from '../clientApi';
+import {selectAnimation} from './AnimationTab/animationTabModule';
+import {reportError} from './errorDialogStackModule';
+/* global dashboard */
+
+// Args: {SerializedAnimationList} animationList
+const SET_INITIAL_ANIMATION_LIST = 'AnimationList/SET_INITIAL_ANIMATION_LIST';
+// Args: {AnimationKey} key, {SerializedAnimation} data
+export const ADD_ANIMATION = 'AnimationList/ADD_ANIMATION';
+// Args: {number} index, {AnimationKey} key, {SerializedAnimation} data
+export const ADD_ANIMATION_AT = 'AnimationList/ADD_ANIMATION_AT';
+// Args: {AnimationKey} key, {SerializedAnimation} data
+export const EDIT_ANIMATION = 'AnimationList/EDIT_ANIMATION';
+// Args: {AnimationKey} key
+const INVALIDATE_ANIMATION = 'AnimationList/INVALIDATE_ANIMATION';
+// Args: {AnimationKey} key, {string} name
+const SET_ANIMATION_NAME = 'AnimationList/SET_ANIMATION_NAME';
+// Args: {AnimationKey} key
+const DELETE_ANIMATION = 'AnimationList/DELETE_ANIMATION';
+// Args: {AnimationKey} key
+const START_LOADING_FROM_SOURCE = 'AnimationList/START_LOADING_FROM_SOURCE';
+// Args: {AnimationKey} key, {Blob} blob, {String} dataURI. Version?
+const DONE_LOADING_FROM_SOURCE = 'AnimationList/DONE_LOADING_FROM_SgOURCE';
+// Args: {AnimationKey} key, {string} version
+const ON_ANIMATION_SAVED = 'AnimationList/ON_ANIMATION_SAVED';
+
+export default combineReducers({
+  list,
+  data
+});
+
+function list(state, action) {
+  state = state || [];
+  switch (action.type) {
+
+    case SET_INITIAL_ANIMATION_LIST:
+      return action.animationList.list;
+
+    case ADD_ANIMATION:
+      return [].concat(
+          state.slice(0),
+          action.key);
+
+    case ADD_ANIMATION_AT:
+      return [].concat(
+          state.slice(0, action.index),
+          action.key,
+          state.slice(action.index));
+
+    case DELETE_ANIMATION:
+      return state.filter(key => key !== action.key);
+
+    default:
+      return state;
+  }
+}
+
+function data(state, action) {
+  state = state || {};
+  var newState;
+  switch (action.type) {
+
+    case SET_INITIAL_ANIMATION_LIST:
+      return action.animationList.data;
+
+    case ADD_ANIMATION:
+    case ADD_ANIMATION_AT:
+    case EDIT_ANIMATION:
+    case INVALIDATE_ANIMATION:
+    case SET_ANIMATION_NAME:
+    case START_LOADING_FROM_SOURCE:
+    case DONE_LOADING_FROM_SOURCE:
+    case ON_ANIMATION_SAVED:
+      newState = Object.assign({}, state);
+      newState[action.key] = datum(newState[action.key], action);
+      return newState;
```

Comment:
I'm not familiar with that syntax; does the `[]` allow me to evaluate an expression to act as the key for an object?